salt/modules/win_event_viewer.py

```diff
@@ -0,0 +1,285 @@
+
+
+def make_event_dict(event):
+    '''
+    info: will make a PyEventLogRecord into a dict
+    :param event: PyEventLogRecord
+    :return: dict
+    '''
+
+    event_dict = {}
+    for event_part in EVENT_PARTS:
+        # get object value and add it to the event dict
+        event_dict[event_part] = getattr(event, event_part[0].upper() + event_part[1:], None)
+
+    # format items
+    event_dict['eventID'] = winerror.HRESULT_CODE(event_dict.get('eventID'))
+    if event_dict.get('sid') is not None:
+        event_dict['sid'] = event_dict.get('sid').GetSidIdentifierAuthority()
+    event_dict['timeGenerated'] = _get_raw_time(event_dict.get('timeGenerated'))
+    event_dict['timeWritten'] = _get_raw_time(event_dict.get('timeWritten'))
+
+    return _change_str_to_bytes(event_dict)
+
+
+def _get_event_handler(log_name, target_computer=None):
+    '''
+    info: will try to open a PyHANDLE
+    :param log_name: str
+    :param target_computer: None or str
+    :return: PyHANDLE
+    '''
+
+    try:
+        return win32evtlog.OpenEventLog(target_computer, log_name)
+    except pywintypes.error:
+        raise FileNotFoundError('Log "{0}" of "{1}" can not be found or access was denied!'.format(log_name,
+                                                                                                   target_computer))
+
+
+def _close_event_handler(handler):
+    '''
+    info: will close the event handler
+    :param handler: PyHANDLE
+    :return:
+    '''
+
+    win32evtlog.CloseEventLog(handler)
+
+
+def get_event_generator(log_name, target_computer=None, raw=False):
+    '''
+    info: will get all log events one by one
+    warning: events are not in exact order
+    :param log_name: str
+    :param target_computer: None or str
+    :param raw: bool: True: PyEventLogRecord False: dict
+    :return: PyEventLogRecord or dict
+    '''
+
+    handler = _get_event_handler(log_name, target_computer)
+    flags = win32evtlog.EVENTLOG_BACKWARDS_READ | win32evtlog.EVENTLOG_SEQUENTIAL_READ
+    event_count = 0
+
+    while win32evtlog.GetNumberOfEventLogRecords(handler) > event_count:
+        # get list of some of the events
+        events = win32evtlog.ReadEventLog(handler, flags, 0)
+        if not events:
+            # event log was updated and events are not ready to be given yet
+            # rather than wait just return
+            break
+        for event in events:
+            event_count += 1
+            if raw:
+                yield event
+            else:
+                yield make_event_dict(event)
+    _close_event_handler(handler)
+
+
+def get_events(log_name, target_computer=None, raw=False):
+    '''
+    info: will make a tuple of log events
+    :param log_name: str
+    :param target_computer: None or str
+    :param raw: bool: True: PyEventLogRecord False: dict
+    :return: tuple
+    '''
+
+    return tuple(get_event_generator(log_name, target_computer, raw))
```

Comment:
I'm not sure that I understand this function - shouldn't we be returning a dictionary from module functions? :suspect:

Reply:
It gets all events from the generator... I'll be using it in the state module

Reply:
That makes sense. Should it need to be `_get_events`?

Reply:
This works
salt-call --local win_event_viewer.get_events System
This does not work
salt-call --local win_event_viewer.get_event_generator System

I think it would be good to have this open to users. They might want to tuple of every event in the log

Reply:
Ah. Yeah, I think that's reasonable - I think I was thinking of states